app/gui2/src/util/range.ts

```diff
@@ -37,3 +34,124 @@ export function* allRanges(
     yield { start: lastEndIndex, end, isMatch: false }
   }
 }
+
+export class Range {
+  constructor(
+    readonly start: number,
+    readonly end: number,
+  ) {}
+
+  /** Create the smallest possible {@link Range} that contains both {@link Range}s.
+   * It is not necessary for the two {@link Range}s to overlap. */
+  merge(other: Range): Range {
+    return new Range(Math.min(this.start, other.start), Math.max(this.end, other.end))
+  }
+
+  exclude(other: Range): Range[] {
+    if (this.start < other.start) {
+      const before = new Range(this.start, other.start)
+      if (this.end > other.end) return [before, new Range(other.end, this.end)]
+      else return [before]
+    } else if (this.end > other.end) return [new Range(other.end, this.end)]
+    else return []
+  }
+
+  intersects(other: Range) {
+    return this.start < other.end && this.end > other.start
+  }
+
+  expand(by: number) {
+    return new Range(this.start - by, this.end + by)
+  }
+}
+
+/** A sorted array of non-intersecting ranges. */
+export class MultiRange {
+  // It is fine for this to be JS-`#private` as it is re-exposed as a public member anyway.
+  readonly #ranges: Range[] = []
+  // This MUST be private, otherwise a consumer may mutate it so that it is no longer sorted or
+  // non-intersecting.
+  readonly ranges: readonly Range[] = this.#ranges
+  constructor() {}
+
+  clear() {
+    this.#ranges.splice(0, this.#ranges.length)
+  }
+
+  insert(range: Range, effectiveRange = range) {
+    const start = binarySearch(this.#ranges, (r) => r.end >= effectiveRange.start)
+    const end = binarySearch(this.#ranges, (r) => r.start > effectiveRange.end, start)
+    let finalRange = range
+    if (end !== start) {
+      const startRange = this.#ranges[start]
+      if (startRange) finalRange = finalRange.merge(startRange)
+    }
+    if (end - 1 > start) {
+      const endRange = this.#ranges[end - 1]
+      if (endRange) finalRange = finalRange.merge(endRange)
+    }
+    return this.#ranges.splice(start, end - start, finalRange)[0]!
+  }
+
+  remove(range: Range, effectiveRange = range) {
+    const start = binarySearch(this.#ranges, (r) => r.end >= effectiveRange.start)
+    const end = binarySearch(this.#ranges, (r) => r.start > effectiveRange.end, start)
+    const finalRanges: Range[] = []
+    if (end !== start) {
+      const startRange = this.#ranges[start]
+      if (startRange) finalRanges.push(...startRange.exclude(range))
+    }
+    if (end - 1 > start) {
+      const endRange = this.#ranges[end - 1]
+      if (endRange) finalRanges.push(...endRange.exclude(range))
+    }
+    return this.#ranges.splice(start, end - start, ...finalRanges)
+  }
+}
+
+if (import.meta.vitest) {
```

Comment:
Could you move that test suite to a dedicated test file? I found that using the in-source testing often breaks external tools (such as vitest vscode extension), so I'd like to migrate away from it where possible.